### PR TITLE
Bump `rng_salt` version to `v0.40.0`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

With v0.40.0 of Pennylane out, we are safe to bump the `rng_salt` version.

**Benefits:** 

This alters the RNG salt of our test suite, ensuring that tests from last release weren't passing due to some magic salt.

**Description of the change:**

This involved modifying some tests but increasing their tolerances.  This was discussed with the core team to ensure we aren't accidently hiding any problems.

[sc-82812]

### CODE REVIEW

**Comment by astralcai** [comment]
Could you also do a search of all `local_salt` that's currently in the test suite and remove the ones that are not necessary for the tests to pass anymore?

**Comment by andrijapau** [comment]
@astralcai, I removed the `local_salt` that don't need to be there. However, lots of new tests are failing as you can see from the CI.

**Comment by astralcai** [comment]
> @astralcai, I removed the `local_salt` that don't need to be there. However, lots of new tests are failing as you can see from the CI.

I guess now is a good time to briefly investigate if these tests are failing because of an unfortunately bad seed or if they've been passing because of a magically good seed. If it's the former, you can add a local salt to make them pass.

File: tests/devices/default_qubit/test_default_qubit.py [hunk 1]
```diff
@@ -1505,8 +1505,9 @@ def test_hamiltonian_expval(self, max_workers, seed):
         res = dev.execute(qs)
 
         expected = 0.8 * np.cos(x) + 0.5 * np.real(np.exp(y * 1j)) * np.sin(x)
-        assert np.allclose(res, expected, atol=0.01)
+        assert np.allclose(res, expected, atol=2e-2)
```
Comment (JerryChen97) [inline tests/devices/default_qubit/test_default_qubit.py:1508]:
```suggestion
        assert np.allclose(res, expected, atol=0.02)
```

**Comment by JerryChen97** [comment]
Seems this PR scope extended a bit to proper adjustment to those stochastic tests right?

File: tests/gradients/core/test_pulse_gradient.py [hunk 2]
```diff
@@ -1268,7 +1268,7 @@ def qnode(params_0, params_1):
         res = fn(qml.execute(tapes, dev, None))
         exp_grad = jax.grad(qnode, argnums=(0, 1))(params_0, params_1)
         exp_grad = exp_grad[0] + exp_grad[1]
-        assert all(qml.math.allclose(r, e, rtol=0.4) for r, e in zip(res, exp_grad))
+        assert all(qml.math.allclose(r, e, rtol=0.53) for r, e in zip(res, exp_grad))
```
Comment (JerryChen97) [inline tests/gradients/core/test_pulse_gradient.py:1271]:
```suggestion
        assert all(qml.math.allclose(r, e, rtol=0.5, atol=0.1) for r, e in zip(res, exp_grad))
```
Can't see what should be the expected values here. So I guess at some point at least one `e` must have become an extremely small value that made a 0.4 `rtol` not enough. So I would like to try an extra `atol` here.

File: tests/measurements/test_expval.py [hunk 3]
```diff
@@ -188,7 +187,7 @@ def circuit():
 
         res = circuit()
         expected = [0.5, 0.5] if isinstance(shots, list) else 0.5
-        assert np.allclose(res, expected, atol=0.02, rtol=0.02)
+        assert np.allclose(res, expected, atol=0.02, rtol=0.035)
```
Comment (JerryChen97) [inline tests/measurements/test_expval.py:190]:
```suggestion
        assert np.allclose(res, expected, atol=0.02, rtol=0.04)
```
`expected=0.5` so `atol` is already 4%. It's fine to just lift `rtol` to same.

**Review by JerryChen97 — APPROVED** [review]
Thanks a lot for the effort! Though I do think it's better to adjust those fine-tuned params to one-digit values, non-blocking suggestions.

**Comment by JerryChen97** [comment]
Lol now I see why
```
=========================== short test summary info ============================
FAILED tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py::TestBroadcastingPRNG::test_nonsample_measure_shot_vector[measurement0-expected0-shots1] - assert False
 +  where False = <function allclose at 0x7f3e328c4cf0>(array([ 0.4959,  1.    , -0.0204]), array([0.5, 1. , 0. ]), atol=0.02, rtol=0.1)
 +    where <function allclose at 0x7f3e328c4cf0> = np.allclose
FAILED tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py::TestBroadcastingPRNG::test_nonsample_measure_shot_vector[measurement0-expected0-shots2] - assert False
 +  where False = <function allclose at 0x7f3e328c4cf0>(array([ 0.49945,  1.     , -0.0216 ]), array([0.5, 1. , 0. ]), atol=0.02, rtol=0.1)
 +    where <function allclose at 0x7f3e328c4cf0> = np.allclose
= 2 failed, 2292 passed, 128 skipped, 214 xfailed, 120 xpassed, 69 warnings in 525.93s (0:08:45) =
Error: Process completed with exit code 1.```

File: tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py [hunk 4]
```diff
@@ -640,7 +640,7 @@ def test_nonsample_measure_shot_vector(self, mocker, shots, measurement, expecte
 
         for r in res:
             assert r.shape == expected.shape
-            assert np.allclose(r, expected, atol=0.02)
+            assert np.allclose(r, expected, atol=0.02, rtol=0.1)
```
Comment (JerryChen97) [inline tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py:643]:
```suggestion
            assert np.allclose(r, expected, atol=0.03)
```

**Comment by JerryChen97** [comment]
> Lol now I see why
> 
> ```
> =========================== short test summary info ============================
> FAILED tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py::TestBroadcastingPRNG::test_nonsample_measure_shot_vector[measurement0-expected0-shots1] - assert False
>  +  where False = <function allclose at 0x7f3e328c4cf0>(array([ 0.4959,  1.    , -0.0204]), array([0.5, 1. , 0. ]), atol=0.02, rtol=0.1)
>  +    where <function allclose at 0x7f3e328c4cf0> = np.allclose
> FAILED tests/devices/qutrit_mixed/test_qutrit_mixed_sampling.py::TestBroadcastingPRNG::test_nonsample_measure_shot_vector[measurement0-expected0-shots2] - assert False
>  +  where False = <function allclose at 0x7f3e328c4cf0>(array([ 0.49945,  1.     , -0.0216 ]), array([0.5, 1. , 0. ]), atol=0.02, rtol=0.1)
>  +    where <function allclose at 0x7f3e328c4cf0> = np.allclose
> = 2 failed, 2292 passed, 128 skipped, 214 xfailed, 120 xpassed, 69 warnings in 525.93s (0:08:45) =
> Error: Process completed with exit code 1.```
> ```

We better directly remove numeric validation for this random `GellMann` expval though. Based on the context, this test suite should only care about the broadcasting pipeline, so only the shape should be checked. Numeric correctness of expectations of all kinds of supported ops should be handled by other specialized test suites.

Also the original testing logic is faulty, it's hard to see the three particularly picked states behave the same way under sampling; at least they are not symmetric under `np.allclose`. Therefore even for its own sake it's better to get numeric validation either removed or refactored.

File: tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py [hunk 5]
```diff
@@ -1370,9 +1370,10 @@ def test_non_involutory_variance_multi_param(self, broadcast, seed):
             assert gradF[0] == pytest.approx(expected, abs=2)
             assert qml.math.allclose(gradF[1], expected, atol=1.5)
 
+    @pytest.mark.local_salt(1)
     def test_involutory_and_noninvolutory_variance_single_param(self, broadcast, seed):
-        """Tests a qubit Hermitian observable that is not involutory alongside
-        an involutory observable when there's a single trainable parameter."""
+        """Tests a qubit Hermitian observable that is not involuntary alongside
+        an involuntary observable when there's a single trainable parameter."""
```
Comment (JerryChen97) [inline tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py:1376]: Why involuntary here?

Comment (andrijapau) [inline tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py:1376]:
```suggestion
        """Tests a qubit Hermitian observable that is not involutary alongside
        an involutary observable when there's a single trainable parameter."""
```

File: tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py [hunk 6]
```diff
@@ -1370,9 +1370,10 @@ def test_non_involutory_variance_multi_param(self, broadcast, seed):
             assert gradF[0] == pytest.approx(expected, abs=2)
             assert qml.math.allclose(gradF[1], expected, atol=1.5)
 
+    @pytest.mark.local_salt(1)
     def test_involutory_and_noninvolutory_variance_single_param(self, broadcast, seed):
-        """Tests a qubit Hermitian observable that is not involutory alongside
-        an involutory observable when there's a single trainable parameter."""
+        """Tests a qubit Hermitian observable that is not involutary alongside
+        an involutary observable when there's a single trainable parameter."""
```
Comment (andrijapau) [inline tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py:1376]:
```suggestion
        """Tests a qubit Hermitian observable that is not involutory alongside
        an involutory observable when there's a single trainable parameter."""
```

**Review by JerryChen97 — APPROVED** [review]

File: tests/workflow/interfaces/qnode/test_jax_jit_qnode.py
Comment (andrijapau) [inline tests/workflow/interfaces/qnode/test_jax_jit_qnode.py:1]: Updated to use theoretical shots tolerances with `rtol` where appropriate to account for any fluctuations?

**Review by andrijapau — COMMENTED** [review]
Comments to go through with team.

Edit: Went through them with team on Feb 27 2025.

File: tests/devices/default_qubit/test_default_qubit.py
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/devices/default_qubit/test_default_qubit.py:1]: After discussion with the team, it is okay to boost the tolerance here.

File: tests/devices/qubit/test_sampling.py
Comment (andrijapau) [inline tests/devices/qubit/test_sampling.py:1]: Should just be testing shapes but we have an `allclose` just incase. Therefore, it's okay to increase the threshold.

File: tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py:1]: After discussion with the team, we are okay with this change.

File: tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py:1]: Team agrees we're just going to deal with it lol.

File: tests/measurements/test_expval.py
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/measurements/test_expval.py:1]: Going to ignore this for now and just re-word the test to be an integration test. This would involve making sure that the projector works with all of the methods in `ExpectationMP` which is beyond the scope of this PR.

File: tests/gradients/core/test_pulse_gradient.py
[the remark this answers is not in the document]
Comment (andrijapau) [reply tests/gradients/core/test_pulse_gradient.py:1]: Nevermind, a lot of other tests use `jax.grad` as well. Will move forward by switching to `atol` since the values are near zero.
